src/Symfony/Component/ObjectMapper/ObjectMapper.php

```diff
@@ -0,0 +1,226 @@
+        }
+
+        $objectMap[$object] = $mapped;
+
+        $arguments = [];
+        $constructor = $toRefl->getConstructor();
+        foreach ($constructor?->getParameters() ?? [] as $parameter) {
+            $parameterName = $parameter->getName();
+            $property = $toRefl->getProperty($parameterName);
+
+            // The mapped class was probably instantiated in a transform we can't write a readonly property
+            if ($property->isReadOnly() && ($property->isInitialized($mapped) && $property->getValue($mapped))) {
+                continue;
+            }
+
+            $arguments[$parameterName] = $parameter->isDefaultValueAvailable() ? $parameter->getDefaultValue() : null;
+        }
+
+        foreach ($refl->getProperties() as $property) {
+            if ($property->isStatic()) {
+                continue;
+            }
+
+            $propertyName = $property->getName();
+
+            $map = null;
+            foreach ($property->getAttributes(Map::class) as $attr) {
+                $map = $attr->newInstance();
+                $if = $map->if;
+
+                if (false === $if) {
+                    continue 2;
+                }
+
+                if ($if && ($fn = $this->getCallable($if)) && !$this->call($fn, null, $object)) {
+                    continue 2;
+                }
+
+                break;
+            }
+
+            $mapToProperty = $map?->to ?? $propertyName;
+            if (!$mapToProperty || !$toRefl->hasProperty($mapToProperty)) {
+                continue;
+            }
+
+            $value = $this->propertyAccessor ? $this->propertyAccessor->getValue($object, $propertyName) : $object->{$propertyName};
+            if ($map && $map->transform) {
+                $value = $this->applyTransforms($map, $value, $object);
+            }
+
+            if (
+                \is_object($value)
+                && ($innerMetadata = $this->metadataFactory->create($value))
+                && ($mapTo = $this->getMapTo($innerMetadata, $value, $object))
+                && (\is_string($mapTo->to) && class_exists($mapTo->to))
+            ) {
+                $value = $this->applyTransforms($mapTo, $value, $object);
+
+                if ($value === $object) {
+                    $value = $mapped;
+                } elseif ($objectMap->contains($value)) {
+                    $value = $objectMap[$value];
+                } else {
+                    $value = $this->map($value, $mapTo->to);
+                }
+            }
+
+            if (\array_key_exists($mapToProperty, $arguments)) {
+                $arguments[$mapToProperty] = $value;
+            } else {
+                $this->propertyAccessor ? $this->propertyAccessor->setValue($mapped, $mapToProperty, $value) : ($mapped->{$mapToProperty} = $value);
+            }
+        }
+
+        if ($arguments) {
+            $constructor?->invokeArgs($mapped, $arguments);
```

Comment:
From what i understand the constructor is called after setting other properties, is this wanted ? I fear that this behavior may confuse some users, i believe it should be better to call it in first and set properties afterwards ?

Reply:
Also i think the constructor should not be called if an existing object was passed to the map function

Reply:
Yes this is wanted to work with readonly properties.

Reply:
But assuming a constructor set an initial value to a property this will override the previous setter no ?

Reply:
Mhh indeed, should I remove the constructor hack alltogether and try to use property accessor to make that work?

Reply:
Constructor usage might be helpful for readonly classes.

Reply:
I think you should keep the constructor, but do the call to the property acessor after this call, it's only a matter of order